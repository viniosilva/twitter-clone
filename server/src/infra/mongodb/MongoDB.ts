import { Connection, createConnection } from 'mongoose';

export default class MongoDB {
  readonly conn: Connection;

  constructor(url: string, dbName: string) {
    this.conn = createConnection(url, {
      dbName,
      useNewUrlParser: true,
      useUnifiedTopology: true,
    });
  }

  isConnected(): boolean {
    return [1, 2].includes(this.conn.readyState);
  }

  async disconnect(): Promise<void> {
    await this.conn.close();
  }
}
