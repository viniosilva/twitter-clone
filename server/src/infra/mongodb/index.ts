import { mongodbConfig } from '../../AppConfig';
import MongoDB from './MongoDB';

export default new MongoDB(mongodbConfig.url, mongodbConfig.dbName);
