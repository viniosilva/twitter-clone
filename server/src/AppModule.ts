import { Module } from '@nestjs/common';
import { LoggerModule } from 'nestjs-pino';
import { httpLogger } from './infra/logger';
import AuthModule from './presentation/auth/AuthModule';
import HealthModule from './presentation/health/HealthModule';
import TweetModule from './presentation/tweet/TweetModule';

@Module({
  imports: [LoggerModule.forRoot({ pinoHttp: httpLogger }), HealthModule, AuthModule, TweetModule],
})
export default class AppModule {}
