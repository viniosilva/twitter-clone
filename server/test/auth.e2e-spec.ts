import { Test, TestingModule } from '@nestjs/testing';
import { INestApplication } from '@nestjs/common';
import request, { Response } from 'supertest';
import AppModule from '../src/AppModule';
import mongodb from '../src/infra/mongodb';
import UserModel from '../src/domain/user/UserModel';

describe('AuthController (e2e)', () => {
  const registerPath = '/api/auth/register';
  const loginPath = '/api/auth/login';
  const tweetPath = '/api/tweets';

  let app: INestApplication;

  beforeEach(async () => {
    const moduleFixture: TestingModule = await Test.createTestingModule({
      imports: [AppModule],
    }).compile();

    app = moduleFixture.createNestApplication();
    await app.init();
  });

  afterEach(async () => {
    await UserModel.deleteMany();
  });

  afterAll(async () => {
    await mongodb.disconnect();
  });

  describe('POST /api/auth/register', () => {
    it('should return the user', () => {
      const payload = { email: 'test@test.com', password: 'secret' };
      return request(app.getHttpServer())
        .post(registerPath)
        .send(payload)
        .expect(201)
        .expect((res: Response) => {
          expect(res.body).toMatchSnapshot();
        });
    });

    it('should throw conflict exception', async () => {
      ``;
      const payload = { email: 'test@test.com', password: 'secret' };
      await request(app.getHttpServer()).post(registerPath).send(payload);

      return request(app.getHttpServer())
        .post(registerPath)
        .send(payload)
        .expect(409)
        .expect((res: Response) => {
          expect(res.body).toMatchSnapshot();
        });
    });

    it('should throw bad request exception', () => {
      const payload = { email: 'test', password: '1' };
      return request(app.getHttpServer())
        .post(registerPath)
        .send(payload)
        .expect(400)
        .expect((res: Response) => {
          expect(res.body).toMatchSnapshot();
        });
    });
  });

  describe('POST /api/auth/login', () => {
    const payload = { email: 'test@test.com', password: 'secret' };

    beforeEach(() => {
      return request(app.getHttpServer()).post(registerPath).send(payload);
    });

    it('should return successfully', async () => {
      return request(app.getHttpServer())
        .post(loginPath)
        .send(payload)
        .expect(200)
        .expect((res: Response) => {
          expect(res.body).toMatchSnapshot({ token: expect.any(String) });
        });
    });

    it('should throw user not found exception', async () => {
      const payload = { email: 'notfound@test.com', password: 'secret' };

      return request(app.getHttpServer())
        .post(loginPath)
        .send(payload)
        .expect(404)
        .expect((res: Response) => {
          expect(res.body).toMatchSnapshot();
        });
    });

    it('should throw bad request exception', () => {
      const payload = { email: 'test', password: '1' };
      return request(app.getHttpServer())
        .post(loginPath)
        .send(payload)
        .expect(400)
        .expect((res: Response) => {
          expect(res.body).toMatchSnapshot();
        });
    });
  });

  describe('POST /api/tweets', () => {
    it('should throw forbiden exception when authorization is empty', () => {
      const payload = { content: 'TEST' };
      return request(app.getHttpServer())
        .post(tweetPath)
        .send(payload)
        .expect(403)
        .expect((res: Response) => {
          expect(res.body).toMatchSnapshot();
        });
    });
  });
});
