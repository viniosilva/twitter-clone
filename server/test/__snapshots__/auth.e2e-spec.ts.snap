// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuthController (e2e) POST /api/auth/login should return successfully 1`] = `
Object {
  "token": Any<String>,
}
`;

exports[`AuthController (e2e) POST /api/auth/login should throw bad request exception 1`] = `
Object {
  "error": "Bad Request",
  "message": Array [
    "email must be an email",
    "password must be longer than or equal to 6 characters",
  ],
  "statusCode": 400,
}
`;

exports[`AuthController (e2e) POST /api/auth/login should throw user not found exception 1`] = `
Object {
  "error": "Not Found",
  "message": "User not found",
  "statusCode": 404,
}
`;

exports[`AuthController (e2e) POST /api/auth/register should return the user 1`] = `
Object {
  "email": "test@test.com",
}
`;

exports[`AuthController (e2e) POST /api/auth/register should throw bad request exception 1`] = `
Object {
  "error": "Bad Request",
  "message": Array [
    "email must be an email",
    "password must be longer than or equal to 6 characters",
  ],
  "statusCode": 400,
}
`;

exports[`AuthController (e2e) POST /api/auth/register should throw conflict exception 1`] = `
Object {
  "error": "Conflict",
  "message": "User already exists",
  "statusCode": 409,
}
`;

exports[`AuthController (e2e) POST /api/tweets should throw forbiden exception when authorization is empty 1`] = `
Object {
  "error": "Forbidden",
  "message": "Forbidden resource",
  "statusCode": 403,
}
`;
